[interp] handle box + brtrue/brfalse and box+isinst+unbox.any patterns up front

lambdageek opened the pull request:
In cases where we don't do cprop and deadce (for example if `mono_interp_opt` is 0 because we're debugging) don't emit a box_vt opcode before a branch.  Instead just emit a constant true

Fixes https://github.com/dotnet/runtime/issues/102988

Also reverts the changes from https://github.com/dotnet/runtime/pull/102998

lambdageek committed:
[interp] handle box + brtrue/brfalse pattern for byreflike types

In cases where we don't do cprop and deadce (for example if
~mono_interp_opt~ is 0 because we're debugging) don't emit a box_vt
opcode before a branch.  Instead just emit a constant true

Fixes https://github.com/dotnet/runtime/issues/102988

lambdageek committed:
Revert "Revert uses of Unsafe.BitCast with spans to unblock mono (#102998)"

This reverts commit c286a8efb3c0d534557dde476d5b54b15c0aeaed.

BrzVlad commented:
This means that the previous approach of retrying with optimization enabled is useless and should be removed. We should handle all IL patterns like this.

Interp changes LGTM.

lambdageek committed:
remove the box byreflike retry code

we handle box byreflike code patterns directly now

lambdageek committed:
handle box;isinst;{brtrue|brfalse|unbox.any} IL patterns

lambdageek committed:
remove the mismatch case in the interp.  throw IPE at execution time

lambdageek committed:
Add test locking down mismatched box;isinst;unbox.any pattern

lambdageek commented:
@BrzVlad PTAL. I updated the interpreter to handle the other patterns in the unoptimized pass.

@AaronRobinsonMSFT I added a new test locking down the behavior for `box S1 ; isinst S2 ; unbox.any S2` - CoreCLR and Mono mini both throw `InvalidProgramException` which seems reasonable - as we discussed Roslyn seems to only ever emit this sequence in `box S1 ; isinst S2; brtrue/brfalse` branches where the types match. 

@stephentoub this PR also reverts your previous revert

AaronRobinsonMSFT commented on src/tests/Loader/classloader/generics/ByRefLike/Validate.cs:
```diff
@@ -81,6 +91,7 @@ public static void Validate_InvalidOpCode_Scenarios()
 
         // Test that explicitly tries to box a ByRefLike type.
         Assert.Throws<InvalidProgramException>(() => { Exec.BoxAsObject(); });
+
```
```suggestion
```

lambdageek committed:
Update src/tests/Loader/classloader/generics/ByRefLike/Validate.cs

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

AaronRobinsonMSFT commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -4827,6 +4836,91 @@ interp_handle_box_patterns (TransformData *td, MonoClass *box_class, const unsig
+			interp_add_ins (td, MINT_LDC_I4_S);
+			td->last_ins->data[0] = (guint16) 1;
+			push_simple_type (td, STACK_TYPE_I4);
+			interp_ins_set_dreg (td->last_ins, td->sp [-1].var);
```
Is the `[-1]` a comment mechanic in the intepreter? is there a macro or some other abstraction over that? I'm assuming the `sp` is "stack pointer", so perhaps that is the previous slot?

lambdageek replied:
Yea, I'm not sure why we do it this way.  But the general pattern is to decrement SP by the number of operands that an opcode expects (so now they're at `sp[0]` ... `sp[k-1]`) and use those `var`s, then do a push for the result(s) (which creates new vars) and then use `sp[-1]` to refer to the result.
